### DISCUSSION TITLE
How to create variables that are accessible from anywhere in my experiment?

### DISCUSSION THREAD

**Discussion by ctreffe** [ctreffe/alfred#148] [Alfred Q&A] [answered]
Similar to the basic concept of a _global variable_, I would like to create variables and store them in a place that is accessible from anywhere in my experiment. For example, the variables should be accessible from inside any hook method in my experiment, or even from within methods in custom classes I create. 

Is there a way to do this in alfred? Also, how can I decide whether these variables are added to my experimental data or not?

**Comment by jobrachem** [comment, marked as answer]
I see two options:

1. Use a **global variable**. This is a viable option, if the value of the variable is *constant* within an experiment session and across experiment sessions. A global variable will *not* be saved in the experiment data. Note that constant, global variables should be written in CAPSLOCK.
2. Use the **`ExperimentSession.adata`** ([docs](https://alfredo3.psych.bio.uni-goettingen.de/docs/generated/alfred3.experiment.ExperimentSession.adata.html)) attribute, which is a dictionary for additional data. You can fill it, for example in the setup hook. The values in the adata dictionary are session-specific and can be savely changed. They will automatically be added to the main experiment data. 

As a side note: You can always manually add values to the experiment data set by adding a `alfred3.Value` element to a page ([docs](https://alfredo3.psych.bio.uni-goettingen.de/docs/generated/alfred3.element.misc.Value.html#alfred3.element.misc.Value))

### Example for global variable

```python
import alfred3 as al
exp = al.Experiment()

MY_VARIABLE = 10

@exp.member
class Demo(al.Page):
    def on_exp_access(self):
        self += al.Text(f"The global value is {MY_VARIABLE}")
```

### Example for additional data

```python
import alfred3 as al
exp = al.Experiment()

@exp.setup
def setup(exp):
    exp.adata["my_variable"] = 10

@exp.member
class Demo(al.Page):
    def on_exp_access(self):
        self += al.Text(f"The value of my variable is {self.exp.adata['my_variable']}")
```
